%%
% Setup
clc
close
clear

% Included tissues in this study:
global Tissues

% addpath(genpath(pwd))

%optional
set(0,'defaultTextInterpreter','none')
set(groot,'defaulttextinterpreter','none');
set(groot, 'DefaultAxesTickLabelInterpreter', 'none')
set(groot, 'defaultLegendInterpreter','none');

% feature astheightlimit 2000

altcolor= [255 255 255;255 204 204; 255 153 153; 255 102 102; 255 51 51;...
    255 0 0; 204 0 0; 152 0 0; 102 0 0;  51 0 0]/255; %shorter 10% = 1 bar

delete clone*.log %delete some files generated by cplex
%% 3.1 Expression data input
load Recon2.v04_Consistent_Model.mat;
% consitent model creation
% A = fastcc_4_rfastcormics(modelR204, 1e-4,0);
% consistent_model = removeRxns(modelR204, modelR204.rxns(setdiff(1:numel(modelR204.rxns),A)));
%%
% setting parameters

load medium_example.mat
load dico_ML.mat
epsilon = 1e-4;
consensus_proportion = 0.9;
already_mapped_tag = 0;
unpenalizedSystems = {'Transport, endoplasmic reticular';
    'Transport, extracellular';
    'Transport, golgi apparatus';
    'Transport, mitochondrial';
    'Transport, peroxisomal';
    'Transport, lysosomal';
    'Transport, nuclear'};
unpenalized = consistent_model.rxns(ismember(consistent_model.subSystems,unpenalizedSystems));
optional_settings.unpenalized = unpenalized;

optional_settings.func = {'DM_atp_c_', 'biomass_reaction'}; % forced reactions

not_medium_constrained = 'EX_tag_hs(e)';% if no constrain is used please remove the field.
optional_settings.not_medium_constrained = not_medium_constrained;

optional_settings.medium = medium_example;% if no medium is used please remove the field.

biomass_rxn = {'biomass_reaction'};

for i = 1 : length(Tissues)
    % Import the gene expression data into Matlab.
    cancerFileName = [Tissues{i}, '_tumor.mat'];
    normalFileName = [Tissues{i}, '_normal.mat'];
%     cd('..\TCGA_Data_Separated_GSE62944\Tumor')   
    load(cancerFileName);
    fpkmCancer = Valuesi;
    rownamesCancer = GeneNames;
    colnamesCancer = SampleIDsi;
    noCancer = length(SampleIDsi);
%     cd .\..
%     cd('..\TCGA_Data_Separated_GSE62944\Normal') 
    load(normalFileName);
%     cd .\..
%     cd('..\Reproducing_the_results')
    fpkm = [fpkmCancer, Valuesi];
    rownamesNormal = GeneNames;
    colnames = [colnamesCancer; SampleIDsi];
    discretized = discretize_FPKM(fpkm, colnames); % no figures
    %%
    % reconstruct the context-specific models
    %
    % consensus models

    [model_cancer, A_final_cancer] = fastcormics_RNAseq(consistent_model, discretized(:,1:noCancer), rownamesCancer, dico, biomass_rxn, already_mapped_tag, consensus_proportion, epsilon, optional_settings);
    [model_control, A_final_control] = fastcormics_RNAseq(consistent_model, discretized(:,noCancer+1:end), rownamesNormal, dico, biomass_rxn, already_mapped_tag, consensus_proportion, epsilon, optional_settings);
    
    models_keep_consensus = zeros(numel(consistent_model.rxns), 2);
    
    models_keep_consensus(A_final_cancer,1) = 1;
    models_keep_consensus(A_final_control,2) = 1;
    save([Tissues{i}, '_CSM.mat'], 'model_cancer', 'model_control', 'models_keep_consensus');
    disp([Tissues{i}, '_CSM.mat', ' is generated.'])
end
